server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/ScanDataSource.java

```diff
@@ -234,32 +239,60 @@ private SortedKeyValueIterator<Key,Value> createIterator() throws IOException {
     }
   }
 
-  @Override
-  public void close(boolean sawErrors) {
-
+  private void returnIterators() {
     if (memIters != null) {
-      tablet.returnMemIterators(memIters);
-      memIters = null;
-      tablet.returnFilesForScan(fileReservationId);
-      fileReservationId = -1;
-    }
-
-    synchronized (tablet) {
-      if (tablet.removeScan(this) == 0) {
-        tablet.notifyAll();
+      log.trace("Returning mem iterators for {}, scanId:{}, fid:{}", tablet.getExtent(),
+          scanDataSourceId, fileReservationId);
+      try {
+        tablet.returnMemIterators(memIters);
+      } finally {
+        memIters = null;
+        log.trace("Returning file iterators for {}, scanId:{}, fid:{}", tablet.getExtent(),
+            scanDataSourceId, fileReservationId);
+        try {
+          tablet.returnFilesForScan(fileReservationId);
+        } catch (IllegalArgumentException | IllegalStateException e) {
+          if (log.isWarnEnabled()) {
+            log.warn("Error Returning file iterators for scan: {}, :{}", scanDataSourceId,
+                e.getMessage());
+            log.warn("{}", ExceptionUtils.getStackTrace(e));
+          }
+          // Continue bubbling the exception up for handling.
+          throw e;
```

Comment:
I've said before that in general, we should choose to either log and handle an exception, or we should choose to pass it through to be logged and handled at another layer. When we do both, this tends to result in multiple log messages for the same exception logged at different layers.

In this case, we're dealing with RTEs, so it's possible we're just throwing it out and not expecting something else to handle it... so this might be the only time it gets logged. But, if that's the case, it's weird that we're just catching 2 specific RTEs and logging, but other possible RTEs that might get thrown.

I'm not sure exactly what the intent here is, so it's hard to make a concrete suggestion, but if we're going to both log and throw, I think the reasoning should be clear why we're doing both.

Reply:
> In this case, we're dealing with RTEs, so it's possible we're just throwing it out and not expecting something else to handle it... so this might be the only time it gets logged. But, if that's the case, it's weird that we're just catching 2 specific RTEs and logging, but other possible RTEs that might get thrown.
> 
I made a change to log any exception rather than just two specific RTEs.


> I'm not sure exactly what the intent here is, so it's hard to make a concrete suggestion, but if we're going to both log and throw, I think the reasoning should be clear why we're doing both.

In this case, the exceptions are being swallowed somewhere in the code. 
The logging would help identity them, while re-throwing the exception ensures that code behavior doesn't change due to the new `catch` block.

I strongly suspect that there are some thrift calls that only recreate Thrift exceptions for the clients and miss other important exception types.